Fixed multi packet send transfer.

DRam86 opened the pull request:
Transport Protocol Connection Management:
 - Changed Request to send to send data[4] to send the number of max packages which can be send in one TP_CM_CTS transfer to 1.
 - Fixed Bytes of TP_CM_CTS data[1] has to be the number of bytes which have to be transfered after this TP_CM_CTS message. Transport Protocol Data Transfer:
 - TP_CM_RTS removed total_number_of_packages transmitted field.
 - TP_CM_CTS fixed package[0] calculation and bytes_send calculation to match RTS data fields.

DanielMartensson commented:
@DRam86 

Have you tried with the examples so they are working too? 

Thank you

DRam86 commented:
It's tested with my own implementation on my target system but i'll test it with the examples.

DRam86 commented:
Is it a deep enough test if the example in the main.c is still working as expected as this is, as far as i understood it, an simple multipacket transfer between 2 ecus over the no processor internal buffer stuctures right?

DanielMartensson commented:
> Is it a deep enough test if the example in the main.c is still working as expected as this is, as far as i understood it, an simple multipacket transfer between 2 ecus over the no processor internal buffer stuctures right?

@DRam86 

Yes. The internal buffer is just s simple queue array. 

If the examples are working perfectly. Then I will accept the pull request

DRam86 commented:
While testing this i ran into an error with the calculation of the number_of_packages_being_transmitted.
I tested this with the Proprietary B.txt example when the messages are transfered e.g. "This is PGN 0xFFFF content       .........." length including '\0' is 44 Bytes . Only 6 packages are transmitted  due to an errorneous calculation of the
```c
		j1939->this_ecu_tp_cm.number_of_packages_being_transmitted = 
  j1939->this_ecu_tp_cm.total_message_size_being_transmitted % 8 > 0 ?
    j1939->this_ecu_tp_cm.total_message_size_being_transmitted / 8 + 1 : 
    j1939->this_ecu_tp_cm.total_message_size_being_transmitted / 8; /* Rounding up */
```
This should be divided by 7 which results in 7 packages where the last package is not completely filled. 
If found this error in several other requests. 
So i created an inline function in the Transport_Layer.h file 
```c
static inline uint8_t SAE_J1939_Transport_Protocol_GetNumberOfPackages(uint16_t total_message_size_being_transmitted) {
	return total_message_size_being_transmitted % 7 > 0 ? total_message_size_being_transmitted / 7 + 1 : total_message_size_being_transmitted / 7; /* Rounding up */
}
```
and replaced all occurences of this calculation with this call to centralize the calculation and reduce the chance to having an error on one implementation.

I also found an error in my fix which i fixed so i'll have to resubmit this pull request should i add the fix to the number of package calculation to this too?

DanielMartensson commented:
> While testing this i ran into an error with the calculation of the number_of_packages_being_transmitted. I tested this with the Proprietary B.txt example when the messages are transfered e.g. "This is PGN 0xFFFF content .........." length including '\0' is 44 Bytes . Only 6 packages are transmitted due to an errorneous calculation of the
> 
> ```c
> 		j1939->this_ecu_tp_cm.number_of_packages_being_transmitted = 
>   j1939->this_ecu_tp_cm.total_message_size_being_transmitted % 8 > 0 ?
>     j1939->this_ecu_tp_cm.total_message_size_being_transmitted / 8 + 1 : 
>     j1939->this_ecu_tp_cm.total_message_size_being_transmitted / 8; /* Rounding up */
> ```
> 
> This should be divided by 7 which results in 7 packages where the last package is not completely filled. If found this error in several other requests. So i created an inline function in the Transport_Layer.h file
> 
> ```c
> static inline uint8_t SAE_J1939_Transport_Protocol_GetNumberOfPackages(uint16_t total_message_size_being_transmitted) {
> 	return total_message_size_being_transmitted % 7 > 0 ? total_message_size_being_transmitted / 7 + 1 : total_message_size_being_transmitted / 7; /* Rounding up */
> }
> ```
> 
> and replaced all occurences of this calculation with this call to centralize the calculation and reduce the chance to having an error on one implementation.
> 
> I also found an error in my fix which i fixed so i'll have to resubmit this pull request should i add the fix to the number of package calculation to this too?

@DRam86 

Wow! Thank you so much!

The library uses C89. If you want the keyword inline to be applied. Use a macro for that. Place that in C89_defines.h header. 

If C89. The defined INLINE should just be a space, else the INLINE should hold the keyword inline.

Thank you.

DanielMartensson commented:
@DRam86 

Is it done? Also the inline macro here? https://github.com/DanielMartensson/Open-SAE-J1939/blob/main/Src/Open_SAE_J1939/C89_Library.h

DRam86 commented:
I don't really have a c89 compiler to check this but i hope this is enough.. i used a prefixed version of the inline macro to avoid errors where multiple libraries define their own INLINE macro and these definitions clash depending on the order of including files. had these problems with several project with vendor provided libraries and open source projects. so using a prefixed library define avoids this.

DanielMartensson commented:
> I don't really have a c89 compiler to check this but i hope this is enough.. i used a prefixed version of the inline macro to avoid errors where multiple libraries define their own INLINE macro and these definitions clash depending on the order of including files. had these problems with several project with vendor provided libraries and open source projects. so using a prefixed library define avoids this.

@DRam86 

Very good! I like it!

Before accepting this pull request.
The documentation must be updated as well. I think this will be a minor change in the documentation.

Do you have a writing program that can save .odt files to .pdf?

Whats need to be done are:

- fields need the have to correct name for CTS and RTS
- fields need to have the correct explanation 
- the last page of the documentation need to have the not of change 

Then it's all done :) thank you very much for solving this issue.

https://github.com/DanielMartensson/Open-SAE-J1939/blob/main/Src/Documentation/Open%20SAE%20J1939.pdf

DRam86 commented:
updated documentation

DanielMartensson commented:
@DRam86 

Thank you. I haven't tried the examples by my self. But if you have tried them and they are working. That's good enough!

Thanks once again.
